Bug: `RuleTester`: cannot read properties of `undefined` (reading `'parse'` )

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

`RuleTester` is no longer able to function in any test case I have been able to find, always throwing a type error before handling the test case.

Looking into the issue, it seems to stem from providedConfig always being undefined. I suspect the issue was introduced by #10147 .

At some point, a property is accessed and we get the TypeError given in the title. In detail, it looks like

```
 FAIL  lint.test.ts > only-ever-use-literal-number-type > valid > type is number
Error: Caught an error while linting
 ❯ Linter.RuleTester.#linter.linter.verify node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:204:17
 ❯ RuleTester.runRuleForItem node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:770:33
 ❯ RuleTester.#testValidTemplate node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:863:25
 ❯ node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:541:40

Caused by: TypeError: Cannot read properties of undefined (reading 'parse')
 ❯ ParserService.parseSync node_modules/eslint/lib/services/parser-service.js:36:33
 ❯ Linter.#flatVerifyWithoutProcessors node_modules/eslint/lib/linter/linter.js:1680:47
 ❯ Linter._verifyWithFlatConfigArrayAndWithoutProcessors node_modules/eslint/lib/linter/linter.js:1992:49
 ❯ Linter._verifyWithFlatConfigArray node_modules/eslint/lib/linter/linter.js:2081:21
 ❯ Linter.verify node_modules/eslint/lib/linter/linter.js:1528:61
 ❯ Linter.RuleTester.#linter.linter.verify node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:202:18
 ❯ RuleTester.runRuleForItem node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:770:33
 ❯ RuleTester.#testValidTemplate node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:863:25
 ❯ node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:541:40 
```

### Reproduction Repository Link

https://github.com/jeremybanka/tseslint-parsing-bug

### Repro Steps

```bash
# 1. clone the repo
git clone https://github.com/jeremybanka/tseslint-parsing-bug.git
# 2. install dependencies
npm install
# 3. run tests
npm test
```

You will see the error shown above.

To verify that it does not occur with the previous version `8.10.0`, simply revert my most recent commit:

```bash
# 1. go back one commit without leaving any staged changes
git reset --hard HEAD^
# 2. reinstall deps
npm install
# 3. rerun tests
npm test
```


### Versions

<!-- Fill in the relevant versions and delete the others -->
| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.11.0` |
| `@typescript-eslint/parser`            | `8.11.0` |
| `@typescript-eslint/rule-tester`       | `8.11.0` |
| `TypeScript`                           | `5.6.3` |
| `ESLint`                               | `9.13.0` |
| `node`                                 | `22.10.0` |

--- a/packages/rule-tester/src/RuleTester.ts
+++ b/packages/rule-tester/src/RuleTester.ts
@@ -166,7 +166,7 @@ function getUnsubstitutedMessagePlaceholders(
 }
 
 export class RuleTester extends TestFramework {
-  readonly #linter: Linter;
+  readonly #lintersByBasePath: Map<string | undefined, Linter>;
   readonly #rules: Record<string, AnyRuleCreateFunction | AnyRuleModule> = {};
   readonly #testerConfig: TesterConfigWithDefaults;
 
@@ -184,31 +184,7 @@ export class RuleTester extends TestFramework {
       rules: { [`${RULE_TESTER_PLUGIN_PREFIX}validate-ast`]: 'error' },
     });
 
-    this.#linter = (() => {
-      const linter = new Linter({
-        configType: 'flat',
-        cwd: this.#testerConfig.languageOptions.parserOptions?.tsconfigRootDir,
-      });
-
-      // This nonsense is a workaround for https://github.com/jestjs/jest/issues/14840
-      // see also https://github.com/typescript-eslint/typescript-eslint/issues/8942
-      //
-      // For some reason rethrowing exceptions skirts around the circular JSON error.
-      const oldVerify = linter.verify.bind(linter);
-      linter.verify = (
-        ...args: Parameters<Linter['verify']>
-      ): ReturnType<Linter['verify']> => {
-        try {
-          return oldVerify(...args);
-        } catch (error) {
-          throw new Error('Caught an error while linting', {
-            cause: error,
-          });
-        }
-      };
-
-      return linter;
-    })();
+    this.#lintersByBasePath = new Map();
 
     // make sure that the parser doesn't hold onto file handles between tests
     // on linux (i.e. our CI env), there can be very a limited number of watch handles available
@@ -220,6 +196,56 @@ export class RuleTester extends TestFramework {
         // ignored on purpose
       }
     });
+  }
+
+  #getLinterForFilename(filename: string | undefined): Linter {
+    let basePath: string | undefined =
+      this.#testerConfig.languageOptions.parserOptions?.tsconfigRootDir;
+    if (filename !== undefined) {
+      // For an absolute path (`/foo.ts`), or a path that steps
+      // up (`../foo.ts`), resolve the path relative to the base
+      // path (using the current working directory if the parser
+      // options did not specify a base path) and use the file's
+      // root as the base path so that the file is under the base
+      // path. For any other path, which would just be a plain
+      // file name (`foo.ts`), don't change the base path.
+      if (filename.startsWith('/') || filename.startsWith('..')) {
+        basePath = path.parse(
+          path.resolve(basePath ?? process.cwd(), filename),
+        ).root;
+      }
+    }
+
+    let linterForBasePath = this.#lintersByBasePath.get(basePath);
+    if (!linterForBasePath) {
+      linterForBasePath = (() => {
+        const linter = new Linter({
+          configType: 'flat',
+          cwd: basePath,
+        });
+
+        // This nonsense is a workaround for https://github.com/jestjs/jest/issues/14840
+        // see also https://github.com/typescript-eslint/typescript-eslint/issues/8942
+        //
+        // For some reason rethrowing exceptions skirts around the circular JSON error.
+        const oldVerify = linter.verify.bind(linter);
+        linter.verify = (
+          ...args: Parameters<Linter['verify']>
+        ): ReturnType<Linter['verify']> => {
+          try {
+            return oldVerify(...args);
+          } catch (error) {
+            throw new Error('Caught an error while linting', {
+              cause: error,
+            });
+          }
+        };
+
+        return linter;
+      })();
+      this.#lintersByBasePath.set(basePath, linterForBasePath);
+    }
+    return linterForBasePath;
   }
 
   /**
@@ -738,6 +764,7 @@ export class RuleTester extends TestFramework {
     let passNumber = 0;
     const outputs: string[] = [];
     const configWithoutCustomKeys = omitCustomConfigProperties(config);
+    const linter = this.#getLinterForFilename(filename);
 
     do {
       passNumber++;
@@ -767,15 +794,7 @@ export class RuleTester extends TestFramework {
             ...configWithoutCustomKeys.linterOptions,
           },
         });
-        messages = this.#linter.verify(
-          code,
-          // ESLint uses an internal FlatConfigArray that extends @humanwhocodes/config-array.
-          Object.assign([], {
-            basePath: filename ? path.parse(filename).root : '',
-            getConfig: () => actualConfig,
-          }),
-          filename,
-        );
+        messages = linter.verify(code, actualConfig, filename);
       } finally {
         SourceCode.prototype.applyInlineConfig = applyInlineConfig;
         SourceCode.prototype.applyLanguageOptions = applyLanguageOptions;
@@ -802,7 +821,7 @@ export class RuleTester extends TestFramework {
       outputs.push(code);
 
       // Verify if autofix makes a syntax error or not.
-      const errorMessageInFix = this.#linter
+      const errorMessageInFix = linter
         .verify(fixedResult.output, configWithoutCustomKeys, filename)
         .find(m => m.fatal);
 
@@ -1255,7 +1274,9 @@ export class RuleTester extends TestFramework {
                       ]).output;
 
                     // Verify if suggestion fix makes a syntax error or not.
-                    const errorMessageInSuggestion = this.#linter
+                    const errorMessageInSuggestion = this.#getLinterForFilename(
+                      undefined,
+                    )
                       .verify(
                         codeWithAppliedSuggestion,
                         omitCustomConfigProperties(result.config),

--- a/packages/rule-tester/src/RuleTester.ts
+++ b/packages/rule-tester/src/RuleTester.ts
@@ -1276,7 +1276,7 @@ export class RuleTester extends TestFramework {
 
                     // Verify if suggestion fix makes a syntax error or not.
                     const errorMessageInSuggestion = this.#getLinterForFilename(
-                      undefined,
+                      item.filename,
                     )
                       .verify(
                         codeWithAppliedSuggestion,

--- a/packages/rule-tester/src/RuleTester.ts
+++ b/packages/rule-tester/src/RuleTester.ts
@@ -201,19 +201,20 @@ export class RuleTester extends TestFramework {
   #getLinterForFilename(filename: string | undefined): Linter {
     let basePath: string | undefined =
       this.#testerConfig.languageOptions.parserOptions?.tsconfigRootDir;
-    if (filename !== undefined) {
-      // For an absolute path (`/foo.ts`), or a path that steps
-      // up (`../foo.ts`), resolve the path relative to the base
-      // path (using the current working directory if the parser
-      // options did not specify a base path) and use the file's
-      // root as the base path so that the file is under the base
-      // path. For any other path, which would just be a plain
-      // file name (`foo.ts`), don't change the base path.
-      if (filename.startsWith('/') || filename.startsWith('..')) {
-        basePath = path.parse(
-          path.resolve(basePath ?? process.cwd(), filename),
-        ).root;
-      }
+    // For an absolute path (`/foo.ts`), or a path that steps
+    // up (`../foo.ts`), resolve the path relative to the base
+    // path (using the current working directory if the parser
+    // options did not specify a base path) and use the file's
+    // root as the base path so that the file is under the base
+    // path. For any other path, which would just be a plain
+    // file name (`foo.ts`), don't change the base path.
+    if (
+      filename !== undefined &&
+      (filename.startsWith('/') || filename.startsWith('..'))
+    ) {
+      basePath = path.parse(
+        path.resolve(basePath ?? process.cwd(), filename),
+      ).root;
     }
 
     let linterForBasePath = this.#lintersByBasePath.get(basePath);

--- a/packages/rule-tester/tests/filename.test.ts
+++ b/packages/rule-tester/tests/filename.test.ts
@@ -1,5 +1,7 @@
 /* eslint-disable perfectionist/sort-objects */
-import { ESLintUtils } from '@typescript-eslint/utils';
+import type { TSESLint } from '@typescript-eslint/utils';
+
+import { AST_NODE_TYPES, ESLintUtils } from '@typescript-eslint/utils';
 
 import { RuleTester } from '../src/RuleTester';
 
@@ -10,35 +12,46 @@ const rule = ESLintUtils.RuleCreator.withoutDocs({
     },
     messages: {
       foo: 'It works',
+      createError: 'Create error',
     },
     schema: [],
     type: 'problem',
+    hasSuggestions: true,
   },
   defaultOptions: [],
   create: context => ({
     Program(node): void {
-      context.report({ node, messageId: 'foo' });
+      context.report({
+        node,
+        messageId: 'foo',
+        suggest:
+          node.body.length === 1 &&
+          node.body[0].type === AST_NODE_TYPES.EmptyStatement
+            ? [
+                {
+                  messageId: 'createError',
+                  fix(fixer): TSESLint.RuleFix {
+                    return fixer.replaceText(node, '//');
+                  },
+                },
+              ]
+            : [],
+      });
     },
   }),
 });
 
 describe('rule tester filename', () => {
-  const ruleTester = new RuleTester();
-
-  ruleTester.run('absolute path', rule, {
+  new RuleTester().run('without tsconfigRootDir', rule, {
     invalid: [
       {
+        name: 'absolute path',
         code: '_',
         errors: [{ messageId: 'foo' }],
         filename: '/an-absolute-path/foo.js',
       },
-    ],
-    valid: [],
-  });
-
-  ruleTester.run('relative path', rule, {
-    invalid: [
       {
+        name: 'relative path above project',
         code: '_',
         errors: [{ messageId: 'foo' }],
         filename: '../foo.js',
@@ -47,29 +60,72 @@ describe('rule tester filename', () => {
     valid: [],
   });
 
-  const ruleTesterWithRootDir = new RuleTester({
+  new RuleTester({
     languageOptions: {
       parserOptions: { tsconfigRootDir: '/some/path/that/totally/exists/' },
     },
-  });
-
-  ruleTesterWithRootDir.run('absolute path with root dir', rule, {
+  }).run('with tsconfigRootDir', rule, {
     invalid: [
       {
+        name: 'absolute path',
         code: '_',
         errors: [{ messageId: 'foo' }],
+        filename: '/an-absolute-path/foo.js',
+      },
+      {
+        name: 'relative path above project',
+        code: '_',
+        errors: [{ messageId: 'foo' }],
+        filename: '../foo.js',
+      },
+    ],
+    valid: [],
+  });
+});
+
+describe('rule tester suggestion syntax error checks', () => {
+  new RuleTester().run('verifies suggestion with absolute path', rule, {
+    invalid: [
+      {
+        code: ';',
+        errors: [
+          {
+            messageId: 'foo',
+            suggestions: [{ messageId: 'createError', output: '//' }],
+          },
+        ],
         filename: '/an-absolute-path/foo.js',
       },
     ],
     valid: [],
   });
 
-  ruleTesterWithRootDir.run('relative path with root dir', rule, {
+  new RuleTester().run('verifies suggestion with relative path', rule, {
     invalid: [
       {
-        code: '_',
-        errors: [{ messageId: 'foo' }],
+        code: ';',
+        errors: [
+          {
+            messageId: 'foo',
+            suggestions: [{ messageId: 'createError', output: '//' }],
+          },
+        ],
         filename: '../foo.js',
+      },
+    ],
+    valid: [],
+  });
+
+  new RuleTester().run('verifies suggestion with no path', rule, {
+    invalid: [
+      {
+        code: ';',
+        errors: [
+          {
+            messageId: 'foo',
+            suggestions: [{ messageId: 'createError', output: '//' }],
+          },
+        ],
       },
     ],
     valid: [],

--- a/packages/rule-tester/tests/filename.test.ts
+++ b/packages/rule-tester/tests/filename.test.ts
@@ -1,6 +1,7 @@
 /* eslint-disable perfectionist/sort-objects */
-import { RuleTester } from '@typescript-eslint/rule-tester';
 import { ESLintUtils } from '@typescript-eslint/utils';
+
+import { RuleTester } from '../src/RuleTester';
 
 const rule = ESLintUtils.RuleCreator.withoutDocs({
   meta: {

--- a/packages/rule-tester/tests/filename.test.ts
+++ b/packages/rule-tester/tests/filename.test.ts
@@ -2,8 +2,6 @@
 import { RuleTester } from '@typescript-eslint/rule-tester';
 import { ESLintUtils } from '@typescript-eslint/utils';
 
-const ruleTester = new RuleTester();
-
 const rule = ESLintUtils.RuleCreator.withoutDocs({
   meta: {
     docs: {
@@ -24,6 +22,8 @@ const rule = ESLintUtils.RuleCreator.withoutDocs({
 });
 
 describe('rule tester filename', () => {
+  const ruleTester = new RuleTester();
+
   ruleTester.run('absolute path', rule, {
     invalid: [
       {
@@ -36,6 +36,34 @@ describe('rule tester filename', () => {
   });
 
   ruleTester.run('relative path', rule, {
+    invalid: [
+      {
+        code: '_',
+        errors: [{ messageId: 'foo' }],
+        filename: '../foo.js',
+      },
+    ],
+    valid: [],
+  });
+
+  const ruleTesterWithRootDir = new RuleTester({
+    languageOptions: {
+      parserOptions: { tsconfigRootDir: '/some/path/that/totally/exists/' },
+    },
+  });
+
+  ruleTesterWithRootDir.run('absolute path with root dir', rule, {
+    invalid: [
+      {
+        code: '_',
+        errors: [{ messageId: 'foo' }],
+        filename: '/an-absolute-path/foo.js',
+      },
+    ],
+    valid: [],
+  });
+
+  ruleTesterWithRootDir.run('relative path with root dir', rule, {
     invalid: [
       {
         code: '_',

